sup3r/models/abstract.py

```diff
@@ -761,22 +778,157 @@ def finish_epoch(self, epoch, epochs, t0, loss_details,
+
+        Parameters
+        ----------
+        low_res : np.ndarray
+            Real low-resolution data in a 4D or 5D array:
+            (n_observations, spatial_1, spatial_2, features)
+            (n_observations, spatial_1, spatial_2, temporal, features)
+        hi_res_true : np.ndarray
+            Real high-resolution data in a 4D or 5D array:
+            (n_observations, spatial_1, spatial_2, features)
+            (n_observations, spatial_1, spatial_2, temporal, features)
+        training_weights : list
+            A list of layer weights that are to-be-trained based on the
+            current loss weight values.
+        optimizer : tf.keras.optimizers.Optimizer
+            Optimizer class to use to update weights. This can be different if
+            you're training just the generator or one of the discriminator
+            models. Defaults to the generator optimizer.
+        multi_gpu : bool
+            Flag to break up the batch for parallel gradient descent
+            calculations on multiple gpus. If True and multiple GPUs are
+            present, each batch from the batch_handler will be divided up
+            between the GPUs and the resulting gradient from each GPU will
+            constitute a single gradient descent step with the nominal learning
+            rate that the model was initialized with.
+        calc_loss_kwargs : dict
+            Kwargs to pass to the self.calc_loss() method
+
+        Returns
+        -------
+        loss_details : dict
+            Namespace of the breakdown of loss components
+        """
+
+        t0 = time.time()
+        if optimizer is None:
+            optimizer = self.optimizer
+
+        if not multi_gpu or len(self.gpu_list) == 1:
+            grad, loss_details = self.get_single_grad(low_res, hi_res_true,
+                                                      training_weights,
+                                                      **calc_loss_kwargs)
+            optimizer.apply_gradients(zip(grad, training_weights))
+            t1 = time.time()
+            logger.debug(f'Finished single gradient descent steps on '
+                         f'{len(self.gpu_list)} GPUs in {(t1 - t0):.3f}s')
+
+        else:
+            futures = []
+            lr_chunks = np.array_split(low_res, len(self.gpu_list))
+            hr_true_chunks = np.array_split(hi_res_true, len(self.gpu_list))
+
+            with ThreadPoolExecutor(max_workers=len(self.gpu_list)) as exe:
+                for i in range(len(self.gpu_list)):
+                    futures.append(exe.submit(self.get_single_grad,
+                                              lr_chunks[i],
+                                              hr_true_chunks[i],
+                                              training_weights,
+                                              device_name=f'/gpu:{i}',
+                                              **calc_loss_kwargs))
+            for i, future in enumerate(futures):
```

Comment:
why don't we need as_completed() here?

Reply:
Also, applying the different gradients like this is better than applying the average of the gradients once?

Reply:
we could use as_completed() but the futures will finish at basically the same time so i was just keeping it simple. And i dont think either is better - averaging with a single weight update might actually be more intuitive, but this is nice and simple.